ludwig/data/prompt.py

```diff
@@ -0,0 +1,176 @@
+
+    if index_name is None:
+        if split_col is None:
+            raise ValueError("split column must be provided if using retrieval")
+        split_col = backend.df_engine.compute(split_col).astype(int)
+
+        # TODO(geoffrey): add support for Dask DataFrames
+        df = pd.DataFrame({name: backend.df_engine.compute(col) for name, col in dataset_cols.items()})
+        df = df[split_col == 0]  # Ensures that the index is only built on the training set
+        retrieval_model.create_dataset_index(df, backend, columns_to_index=[col_name])
+        index_name = f"embedding_index_{uuid.uuid4()}"
+        logger.info(f"Saving index to cache directory '{index_cache_directory}' with name '{index_name}'")
+        retrieval_model.save_index(index_name, cache_directory=index_cache_directory)
+    else:
+        logger.info(f"Loading index from cache directory '{index_cache_directory}' with name '{index_name}'")
+        retrieval_model.load_index(index_name, cache_directory=index_cache_directory)
+    return retrieval_model, index_name
+
+
+def format_input_with_prompt(
+    input_col_name: str,
+    input_col: Series,
+    backend: "Backend",
+    task_str: str,
+    retrieval_model: Optional[RetrievalModel] = None,
+    template: Optional[str] = None,
+) -> Series:
+    """Returns a new Series with the input column data formatted with the prompt.
+
+    A prompt can either be zero-shot or few-shot. A zero-shot prompt is comprised of some (unlabeled) input and a task
+    to be completed given the input. A few-shot prompt additionally includes some dynamically retrieved context, which
+    is retrieved using the `retrieval_model.search` function.
+
+    A template can be provided to customize the prompt. The template must be a string with the following fields:
+        - sample_input: The input sample.
+        - task: The task to be completed given the input.
+        - context: The context retrieved by the `search_fn` function. Only required if `search_fn` is provided.
+    """
+    # determine if this is a few-shot or zero-shot prompt
+    # few-shot prompts require a search function that returns samples from some dataset
+    is_few_shot = retrieval_model is not None
+
+    # if no template is provided, use the default template
+    if template is None:
+        if is_few_shot:
+            template = DEFAULT_FEW_SHOT_PROMPT_TEMPLATE
+        else:
+            template = DEFAULT_ZERO_SHOT_PROMPT_TEMPLATE
+
+    # ensure that the prompt template has all required fields
+    try:
+        _validate_prompt_template(template, is_few_shot=is_few_shot)
+    except ValueError as e:
+        raise ValueError(f"template invalid for {'few-shot' if is_few_shot else 'zero-shot'} prompt: {e}")
+
+    def generate_prompt(series: pd.Series):
+        df = series.to_frame(name=input_col_name)
+
+        if is_few_shot:
+            df["context"] = retrieval_model.search(df, backend, k=3, return_data=True)
```

Comment:
Probably don't want to hard-code k right?

Reply:
Done.